### ISSUE TITLE
connector.findAndConnect() does not work or is undetermined during model building

### ISSUE THREAD

**Issue by aymanhab** [opensim-org/opensim-core#418]
Main issue is that iterators that are utilized by the method are not available until after model construction/initSystem is returned because they assume all components have been added to their parents and piggyback on the recursive calls to build the tree of components. Using the iterators prematurely is problematic since it depends on the order of processing which is impossible to enforce.

We were just lucky to be able to find objects to connect to because the objects we're connecting to were processed first.

**Comment by aseth1** [comment]
What is the case where you are trying to connect to a Component before its root Component has been constructed and has had the opportunity to be connected? The root Component must have `connect()` called on it (which builds the traversal tree). You cannot connect to any Component that lives outside of the tree defined by the root. Why is this not sensible to enforce?

**Comment by aymanhab** [comment]
@aseth1 Here's a use case from testFrames: double_pendulum.osim Mesh geometry needs to connect to Frame rod1_Frame_1 (PhysicalOffsetFrame) which lives under top level model but has not been processed yet so its "connect" attempt fails. In general unless every component connects only to components that appear before it in traversal, this will fail mysteriously. The documentation of getComponentList states clearly that  "All addComponent calls must be done before calling this method on the top model" for this specific reason.

**Comment by aymanhab** [comment]
@aseth1 to show that this is not a new bug introduced in geometry_integration and that we're just lucky not to run into this so-far I recreated the problem in master by just adding a PhysicalOffsetFrame in a double pendulum (with 0 offset) and tried to make a joint use it and it failed the same way throwing an exception that it can't find the Frame (order dependence). While we can temporarily work around this by adding the Frames first it clearly isn't a new bug AFAIK. I can send the modified model file separately be email since I can't attach it here.

**Comment by aseth1** [comment]
I don't doubt there is a real issue. It seems clear that the first thing that has to happen is that the tree is built on the root. This doesn't seem to be happening and it it is running down the list of subcomponents calling connect() on them first which is incorrect.  I see that `initComponentTreeTraversal()` is called inside `connect` and that is clearly too late. Why not have the top level (e.g. Model) call `initComponentTreeTraversal` when it is done its `extendFinalizeFromProperties`. Or you could add it to `Component::finalizeFromProperties` as the last thing a Component does before returning from that method.

**Comment by aymanhab** [comment]
@aseth1 We have been trying to not treat the model as "special" which implies that methods on Model are called on ModelComponent and that goes down recursively. If you want to change that so that the Model's method is special then fine we can do that too but when/how additional components are added to the tree is still unclear as per #421

**Comment by aseth1** [comment]
Then:
>you could add it to Component::finalizeFromProperties as the last thing a Component does before returning from that method.

would be more general and would not have Model as a special case.

**Comment by aymanhab** [comment]
@aseth1 finalizeFromProperties method doesn't take "root". Any suggestions?

**Comment by aseth1** [comment]
Perhaps this a problem. `Component::initComponentTreeTraversal()` takes the root Component as an input argument but only uses it to verify the terminal node. The tree traversal begin with the root, no?

**Comment by aymanhab** [comment]
@aseth1 I think all this is fine and works as expected, since after all addComponent calls are made, and initTraversal is finished then the correct set of Components that need traversal are returned. The issue is trying to use it during the recursive call and before the call on top level model is finished. This can be solved by splitting the current call to connect(root) into multiple parts, one that calls addSubComponent(s), then a call to initTraversal , then a call to connect(root) as it is today. Please let me know if you agree with that solution.

**Comment by aseth1** [comment]
The adding of the subcomponents should be complete by `connect()`. Unless the model has been edited in which it will force `finalizeFromProperties()` in the beginning of `connect()`. Is this the case when you see the problem? It may explain why some components are not in the subcomponents list.

**Comment by aymanhab** [comment]
@aseth1 I'm not sure what you mean by "complete by connect()" is it before connect, during the recursive call to connect or after return. 

I created a branch findAndConnectissue418 where initTraversal is only called from the top model and the iteration finds all components correctly before calling connect but then crashes downstream in many cases e.g. testFrames because of some other calls that reset/wipe components. Please let me know if you have a conjecture.

**Comment by aseth1** [comment]
I mean the subcomponent lists should have been populated by the time `connect()` is invoked on any of the components. This is done in `finalizeFromProperties()` independent and prior to `connect()` except in the case where components have been edited between the two calls, in which case connect() invokes finalizeFromProperties itself and that could be problematic. Is this in fact the case with the connecting to a Frame issue?

**Comment by aymanhab** [comment]
@aseth1 I put a break point in the block of code that invokes finalizeFromPropertie sinside connect and it was invoked on all Geometry objects as well as Joints. I'm not sure what that means since there's no "editing" done as far as I can tell, is there a documentation or description of what constitute an edit and when is it allowed?. Also important to note that all this code executes perfectly fine if you use component.find and fails if you use findAndConnect. Unless there're some other mysterious interactions this suggests the issue has to do with premature call in findAndConnect.

**Comment by aseth1** [comment]
An edit is any change to properties, in which case the `isObjectUpToDateWithProperties()` will return false. That means the the Component has not finalized itself from its properties since the last change or its creation. You could try calling finalizeFromProperties() explicitly before connect() is invoked.

findComponent does not use the iterator interface - it just recursively searches through the subcomponents for a matching name. It doesn't do any type filtering like the iterator supports.
I still think the issue is that the tree is incomplete and the fact that finalizeFromProperties is being invoked in connect may be the culprit since connect is attempting to construct the tree of subcomponents whilst finalize is adding subcomponents.

**Comment by aymanhab** [comment]
@aseth1 in the above scenario finalizeFromProperties() is invoked indeed (on Geometry and Joints) but addComponent is never called.

**Comment by aymanhab** [comment]
@aseth1 Thanks for your help troubleshooting, now the findAndConnect works. I also removed the disconnect call and nothing changed so that's good. There were three test failures:
testFrames: due to iterator invoked after connect() rather than initSystem() which I fixed. 
testJoints which fails due to slave body not found (seems like a violation of the principle that all addComponent calls done before building the traversal tree), 
testComponentInterface: Investigating.
Latest code is on branch findAndConnectissue418

**Comment by aymanhab** [comment]
I reverted the change to the model in testIK and the same failures happen on travis. The change in that model is nunnecessary if we use findAndConnect since it filters by type.

**Comment by aymanhab** [comment]
@aseth1 Per discussion with @sherm1 regarding removing Super:: qualification and the name World::connect() which is unrelated to Component::connect(Component&) it seemed that the reuse of the name in Wold::connect is confusing and any other name would do, so I changed the name to buildComponentTreeAndConnect(), now the code has no improper calls to Super::  and all cases pass. Since the method name is local to the test case we can modify very easily as needed. The branch has been merged to geometry_integration_tabs as you indicated earlier that the use of Super:: was the only hold up.
